### ISSUE TITLE
[css-contain-2] Clarify whether content-visibility: auto subtree elements are visible in accessibility

### ISSUE THREAD

**Issue by vmpstr** [w3c/csswg-drafts#5857]
The [contain spec level 2](https://www.w3.org/TR/css-contain-2/#cv-a11y) says that content-visibility: hidden elements should be omitted from the accessibility tree.

However, it doesn't say whether content-visibility: auto elements should be included or omitted. It only says that the timing of determining whether to include or omit the element should not differ greatly from non-accessibility path.

We believe that the content-visibility: auto elements should be visible in the accessibility. There is a caveat: in order to keep timing the same not process style, we would assume that the off-screen unrendered elements are visible even if they have something like visibility: hidden, until they are brought into view and style is updated.

/cc @alice @chrishtr

**Comment by alice** [comment]
It would be worth hearing from AT users and other experts on this.

My sense is that the actual user impact of exposing `visibility: hidden` and `display: none` elements (including `<script>` and `<style>` elements) inside `content-visibility: auto` areas to AT will be minimal.

Firstly, I'm imagining it will typically be used on blocks which contain only content - scripts and styles would not tend to be in content which should be rendered on-demand.

Secondly, even if they are in that content, they aren't going to be exposed to users of AT while they're not in or near the visible viewport, because AT doesn't typically provide a way to explore off-screen content in depth, since a majority of users of AT can access the visible content in some way.

The only scenario I can immediately think of where it might be an issue for users is AT find-in-page, which could find content inside of offscreen `<script>` or `<style>` elements, cause them to be scrolled into view, and thus cause their styles to be resolved and those elements to be hidden.

I _think_, given the unlikelihood of this actually happening frequently, it's a reasonable trade-off against the other options we could think of, namely:

- resolve style for all `content-visibility: auto` elements if AT APIs are being accessed, or
- the current implementation, which effectively doesn't expose `content-visibility: auto` elements (including headings and landmarks) to AT (since they are exposed as "hidden", analogous to how `aria-hidden` regions are exposed).

I guess a final option might be to _only_ expose nodes matching a safelist of computed roles, but that sounds like a bit of a can of worms potentially.

@cookiecrook any thoughts? Anyone else to tag in?

**Comment by jakearchibald** [comment]
> Firstly, I'm imagining it will typically be used on blocks which contain only content - scripts and styles would not tend to be in content which should be rendered on-demand.

https://www.theguardian.com/uk has a number of non-empty script & style elements in elements that may benefit from `content-visibility: auto`.

> Secondly, even if they are in that content, they aren't going to be exposed to users of AT while they're not in or near the visible viewport, because AT doesn't typically provide a way to explore off-screen content in depth, since a majority of users of AT can access the visible content in some way.

I hadn't appreciated this. This makes me worry less 😄 

> The only scenario I can immediately think of where it might be an issue for users is AT find-in-page

I guess we can't detect when this happens? Otherwise, we could deopt at this point, as we do with regular browser find-in-page.

I can imagine similar cases where an AT user looks at the heading structure of the page, but some headings are inside custom 'tab content' of a non-selected tab. This would cause the heading to become hidden when the browser tries to scroll to the heading.

I guess this wouldn't be an issue if the tab UI used the `hidden` or `aria-hidden` attribute to hide non-selected tabs. But it might be an issue if the site relies on `display: none` to hide content from both visual and AT users.

**Comment by joelanman** [comment]
>AT doesn’t typically provide a way to explore off-screen content in depth

isn't the Rotor in Voiceover an example of this?

https://support.apple.com/en-gb/HT204783

**Comment by flackr** [comment]
Could we expose the currently offscreen top-level content-visibility: auto elements in the accessibility tree in a way that accessibility readers would know that visiting that item would reveal some content? We could suggest that developers put appropriate title attributes on those elements so users would have context.

**Comment by vmpstr** [comment]
I don't know if there is a way in the accessibility language to express the fact this is something that will change if on-screen.

The education of putting the right headings can definitely work, but that is coupled with us hiding content by default which means that developers that don't follow best practices end up hiding more than desired from a11y users.

Alternatively the developer education can come in the form of "use aria labels if you really want to hide content" and show it by default :) (with an opposite caveat is that developers that don't follow best practices reveal more than desired to a11y users)

**Comment by flackr** [comment]
Well my proposed language would just be something that can be revealed if you "activate" it. The fact that it is activated by scrolling it into view is a detail that probably doesn't need to be expressed.

**Comment by JAWS-test** [comment]
Invisible elements (e.g. display:none) should never be included in the accessibility tree.
Offscreen elements should always be included in the accessbility tree (e.g. for element overview and quick navigation with the screen reader).
If there is a conflict between the two requirements for content-visibility: auto, content-visibility: auto must not be used.

**Comment by frivoal** [comment]
I think this case should help us think through this: in documents like social network feeds, with infinite scrolling and loading of content as you scroll, you sort of have 3 classes of elements: those are are on screen, those have have been loaded but are off-screen, those that haven't been added to the document yet, but will be once you scroll further. The later might have been fetched off the network already and be known to javascript only, or maybe they haven't even been loaded yet. Nonetheless, they are part of the conceptual document, and will eventually be visible, both by user who look at the box tree and those who navigate via the accessibility-tree. But until they are actually loaded into the DOM, they are invisible to both. For short, I'll call these off-DOM elements.

While elements with `content-visibility:hidden` (not `auto` as discussed in this issue) are in the DOM, they are closer to off-DOM elements (or `display:none` elements): not searchable, not tabbable, not focusable, etc. So it makes sense for them not to be reachable via the accessibility tree, as is the case for `display: none` elements.

Elements with `content-visibility:auto` are closer to off-screen elements than to off-DOM elements, as they can be searched, focused, tabbed to… So logically, they ought to be accessible through the accessibility tree as well.

The reason it's been suggested that they should be skipped is that applying styles, which is necessary to resolve things like `display:none` and other things that affect the accessibility tree, is expensive, and we'd rather skip that.

But I don't believe this argument holds, and we actually need to do the styling anyway, regardless of the accessibility tree situation: find-in-page also depends on `display:none`. So does tabbing, focusing, or selecting which are explicitely allowed on `content-visibility:auto` elements. So while we might not need to *layout* the `content-visibility:auto` subtree, we do need to style it anyway. and if we're going that far, there's no need to do anything unusual for the accessibility tree.

I think `content-visibility: hidden-matchable`, as proposed in https://github.com/w3c/csswg-drafts/issues/5595, would be the hard case, where it's much more ambiguous as to whether it's closer of off-screen or off-DOM , but this third state hasn't been accepted into css.

**Comment by vmpstr** [comment]
I agree with the above reasoning for the most part. One distinction I'd like to draw is how much of the style we need for all of the cases you mentioned:

1. Find-in-page is an explicit (uncommon) request by the user, it needs enough style to find the next match. This can either be done by forcing the style update on each content-visibility: auto element, trying to find the match, and failing to do that continuing onto the next content-visibility: auto element. (Full disclosure: Chromium forces the update on all content-visibility: auto elements, since we believe that find-in-page requiring an extra bit of rendering is acceptable, but that's an implementation detail not a requirement).
2. tabbing/focussing is also an explicit request by the user, but arguably more common. The way it's done in Chromium is to find the next _potential_ focusable element in the DOM, update its style to ensure that it is indeed not hidden in any way. So in effect we would still update one content-visibility: auto subtree at a time.
3. For accessibility, we have to produce the full accessibility tree to an external AT, such as a screen reader, regardless of how much or how little the user intends to explore it. This means we have to update the style for every content-visibility: auto element, regardless of its distance to the viewport.

So for cases other than accessibility, we style in response to an explicit user action and we are able to only style a subset of content. In accessibility, the mere presence of accessibility tech would cause us to always update style and it has to be done on all content-visibility: auto elements.

As was mentioned in the meeting, this also potentially exposes a11y users via timing attacks. Or is the proposal to always style content-visibility: auto elements, regardless of a11y?

**Comment by frivoal** [comment]
So, if I understand correctly, if we could have something like a lazy accessibility tree, where you can compute things on demand rather than ahead of time, there would be no difference between accessibility and tabbing or search in page, but since we don't, there is a difference. Is that right?

> As was mentioned in the meeting, this also potentially exposes a11y users via timing attacks. Or is the proposal to always style content-visibility: auto elements, regardless of a11y?

That would be my proposal indeed, that said, I am not sure I understand the timing attack. Due to find in page (and tabbing), non-accessibility cases also sometimes need to compute the styles, so a delay caused by style computation would not necessarily be caused by usage of the accessibility tree, so it does not single out accessibility tree users.

> Find-in-page is an explicit (uncommon) request by the user, it needs enough style to find the next match.

Finding just the next one not enough, at least not to support the same behavior as currently. While it is true that the next match is highlighted in a particular way that is distinct from subsequent matches, and that even if you would normally highlight (in a different way) the rest, that doesn't matter because they are hidden anyway, you still need to find the rest:
* All of Safari/Chrome/Firefox display the total number of matches
* Chrome puts colored marks in the scollbar to indicate the places where find-in-page has found matches
* If I understand correctly, when you use macOS's system-wide search (spotlight), it will find currently open pages (and maybe also history? I'm not sure) that contain the relevant terms, including if they are in a hidden-but-findable subtree, and that's powered by webkit indexing these things as it renders them, which means it needs to run styling on those parts as well.

**Comment by vmpstr** [comment]
> So, if I understand correctly, if we could have something like a lazy accessibility tree, where you can compute things on demand rather than ahead of time, there would be no difference between accessibility and tabbing or search in page, but since we don't, there is a difference. Is that right?

That is correct. As you mentioned earlier, if you consider something like a traditional virtual-scroller where the page adds and removes elements in the stream, in part to make sure the scroller has good performance, then it is clear that the accessibility tree cannot include things that are not in the DOM.

Content-visibility: auto attempts to bring similar performance to regular scrollers by skipping rendering automatically when elements are off-screen.

The a11y pass though produces a different view onto the document and hands that view over to a11y technologies which allow the user to explore the document outside of UA. Here, for the most part the visual viewport is not relevant since the user can explore content which would otherwise be off-screen.

If it was possible for us to provide a view into a partial document or a part of a document, with an annotation specifying that accessibility technology should request more of the document if it needs it, then that would solve the problem for sure. Unfortunately I don't think that's possible. We could expose content-visibility: auto unrendered elements as a text node saying "Please bring this into view to reveal contents" or something like that, but I don't think that would acceptable :)

> Finding just the next one not enough, at least not to support the same behavior as currently.

I think the difference is what can be measured by script. In Chrome, the "rest of the matches" other than the first don't block rendering and are done during idle time. But in general you're right that there are times when other activities can be detected by timing. The problem, as I understand it, is that accessibility  would _always_ cause this rendering. That is, If I put something into the DOM that has content-visibility: auto and purposefully heavy style in its subtree, and measure the time it takes to render this, then more likely than not I would be able to tell whether style was processed or not and thus infer that either the user happens to have been searching for something, or that they are using accessibility technologies. 

I would personally be against making content-visibility: auto always process style, since this is a feature that has been designed to skip such work for off-screen content in the first place. Always styling reduces the value of the feature quite a bit. 

There is an argument to be made for making this off-screen content _not_ available to accessibility, since the spec says that this content is not rendered or painted, which would imply that it is similar to visibility: hidden which I believe hides things from screen readers

**Comment by dlibby-** [comment]
> If it was possible for us to provide a view into a partial document or a part of a document, with an annotation specifying that accessibility technology should request more of the document if it needs it, then that would solve the problem for sure.

We have been looking into this more general problem ([see this explainer in WICG](https://github.com/WICG/accessible-loading-and-searching-of-content/blob/main/explainer.md)) and have an [experimental implementation](https://groups.google.com/a/chromium.org/g/blink-dev/c/etQ-LVhY6S4/m/QXG4acPAAwAJ) in Chromium to expose this to ATs on Windows (behind a flag). We're currently working with the Narrator team in Windows to implement experimental support for this new property in order to explore the user experience in practice.

`content-visibility: auto` would then respond naturally to AT actions that cause it to scroll into (or close to) the viewport.

That being said, there probably is something to be said for `content-visibility: auto` working for AT users 'out of the box' without authors needing to add extra ARIA semantics to ensure content is fully accessible.

**Comment by MReschenberg** [comment]
I brought this up at our last a11y team meeting, and the general consensus was exposing offscreen content-visibility:auto content to AT's is important, and if the cost of doing that is causing some hidden content to render into the tree temporarily (until style comp), then that's a better tradeoff than not exposing the content at all. In general, we're fine with the Chrome approach.

Technically, neither case fits with what AT's expect currently. The discussion on the call today seemed to indicate we're looking for a solution that maintains the tree invariants AT's expect. In that case, I think we've identified style computation as a necessary part of that solution. I'd like to flag here that doing style comp for AT's only is something we'd like to avoid because it has the potential to expose whether or not a user is using an AT. In our view, that tradeoff is also worse than rendering all content into the tree. With all that said, I think it might be worthwhile to consider modifying content-visibility such that style comp does happen, but painting doesn't -- that'd still allow UA's to optimise a bit without sacrificing AT experience.

**Comment by JAWS-test** [comment]
> Rossen_: Do we know if JAWS_test is from the JAWS team? The comment they added on May 17

I am not from Freedom Scientific (JAWS). I am an accessibility expert

**Comment by frivoal** [comment]
Another possibility could be to generally include the off-screen content without computing styles, but skipping subtrees rooted in elements that are hidden by default, to reduce false false positives. Basically, everything that's `display:none` in the UA stylesheet.
```css
head, link, meta, script, style, title,
base, basefont, datalist, noembed,
area, param,
rp,
noframes,
template,
input[type="hidden" i],
[hidden],
audio:not([controls]),
dialog:not([open])
```

Checking against this will have a non-zero cost, but should hopefully be a lot cheaper than performing a full styling.
